### PR TITLE
Sharding Detection with multiple IO

### PR DESCRIPTION
**Pull request by samnordmann**
# What

Patch resharding detection routines in `multidevice/utils` to handle expressions with multiple I/O

# Why

As a step toward https://jirasw.nvidia.com/browse/NVFUSER-106, the motivation for this patch is to be able to keep the MatmulOp (and potentially other ops in the future) as a resharding OP, which will undergo a special HostIr lowering (bypassing `ReorderShardedAxisPass`)

### CODE REVIEW

**Comment by samnordmann** [comment]
!test

File: csrc/multidevice/utils.cpp [hunk 1]
```diff
@@ -42,45 +42,37 @@ std::unordered_set<IterDomain*> getShardedIterDomains(TensorView* tv) {
   return sharded_ids;
 }
 
-// Returns whether a IterDomain in a TensorView is the outermost
-// allocated IterDomain in the TensorView.
-bool isOutermostAllocatedId(TensorView* tv, IterDomain* id) {
+// Returns the position where an axis is allocated in a tv. Returns -1 if id is
```
Comment (wujingyue) [inline csrc/multidevice/utils.cpp:45]:
```suggestion
// Returns the position where an axis is allocated in a tv, skipping trivial dimensions (i.e. DID, reduction and broadcast). Returns -1 if id is
```

File: csrc/multidevice/utils.cpp [hunk 2]
```diff
@@ -42,45 +42,37 @@ std::unordered_set<IterDomain*> getShardedIterDomains(TensorView* tv) {
...
+// not in tv's loop domain WAR: today we assume that the loop domain match with
+// the actual allocation, but this will have to change in the future.
+int64_t allocationIndex(TensorView* tv, IterDomain* id) {
+  int64_t index = 0;
   for (auto i : tv->getLoopDomain()) {
```
Comment (wujingyue) [inline csrc/multidevice/utils.cpp:50]:
```suggestion
  for (auto* loop_id : tv->getLoopDomain()) {
```

**Review by wujingyue — APPROVED** [review]

**Comment by wujingyue** [comment]
!test

**Pull request merged**